docs/dev/rfcs/query/0007-semi-structured-data-types.md

```diff
@@ -0,0 +1,106 @@
+# Semi-structured data types design
+
+## Summary
+
+Semi-structured data types are used to represent schemaless data formats, such as JSON, XML, and so on.
+In order to be compatible with [Snowflake's SQL syntax](https://docs.snowflake.com/en/sql-reference/data-types-semistructured.html), we support the following three semi-structured data types:
+
+- `Variant`: A tagged universal type, which can store values of any other type, including `Object` and `Array`.
+- `Object`: Used to represent collections of key-value pairs, where the key is a non-empty string, and the value is a value of `Variant` type.
+- `Array`: Used to represent dense or sparse arrays of arbitrary size, where the index is a non-negative integer (up to 2^31-1), and values are `Variant` types.
+
+Since `Object` and `Array` can be regarded as a type of `Variant`, the following introduction mainly takes `Variant` as an example.
+
+### Data storage format
+
+In order to store the `Variant` type values in the `parquet` format file with schema, we need to do some conversion on the original raw value. We have the following two choices:
+
+#### Store data in one column as JSON or binary JSON-like format
+
+JSON (JavaScript Object Notation) is the most common semi-structured format that can represent arbitrarily complex hierarchical values. It is very suitable for representing this kind of semi-structured data. Data of type `Variant` can be encoded in JSON format and stored as a raw string value.
+The main disadvantage of the JSON format is that each access requires expensive parsing of the raw string, so there are several optimized binary JSON-like formats to improve parsing speed and single key access.
+For example, MongoDB and PostgreSQL use [BSON](https://bsonspec.org/) and [jsonb](https://www.postgresql.org/docs/14/datatype-json.html) respectively to store data in JSON format.
+[UBJSON](https://ubjson.org/) is also a compatible format specification for binary JSON, it can provide universal compatibility, as easy of use as JSON while being faster and more efficient.
+All of these binary JSON formats have better performance, the only problem is they lack a good Rust implementation libraries.
```

Comment:
for further storage layout optimization, I think this paper also provide some insights, https://arxiv.org/pdf/2111.11517.pdf

Reply:
ok, I'll look into this paper